File: ideal-core/src/main/java/cn/idealframework2/event/EventListenerRegistry.java
package cn.idealframework2.event;

import javax.annotation.Nonnull;
import java.util.function.Consumer;

/**
 * @author 宋志宗 on 2022/9/29
 */
public interface EventListenerRegistry {

  /**
   * 监听事件
   *
   * @param name     监听器的全局唯一名称
   * @param clazz    事件类型
   * @param consumer 处理逻辑
   * @param <E>      事件类型
   */
  <E> void register(@Nonnull String name,
                    @Nonnull Class<E> clazz,
                    @Nonnull Consumer<E> consumer);
}

File: ideal-core/src/main/kotlin/cn/idealframework2/event/coroutine/EventListenerRegistry.kt
package cn.idealframework2.event.coroutine

import cn.idealframework2.event.Event
import kotlinx.coroutines.CoroutineScope

/**
 * 事件监听器管理器
 *
 * @author 宋志宗 on 2022/4/6
 */
interface EventListenerRegistry {

  /**
   * 监听事件
   *
   * @param name  监听器名称
   * @param clazz 事件类型
   * @param block 处理逻辑
   * @author 宋志宗 on 2022/4/8
   */
  fun <E : Event> register(
    name: String,
    clazz: Class<E>,
    block: suspend CoroutineScope.(E) -> Unit
  )
}
